src/libcrun/container.c

```diff
@@ -3706,6 +3773,160 @@ libcrun_container_update_from_values (libcrun_context_t *context, const char *id
   return ret;
 }
 
+void
+populate_array_field (const char ***field, const char *array[], size_t num_elements)
+{
+  *field = xmalloc0 ((num_elements + 1) * sizeof (char *));
+  size_t i;
+  for (i = 0; i < num_elements; i++)
+    {
+      (*field)[i] = xstrdup (array[i]);
+    }
+
+  (*field)[i] = NULL; // Terminate the array with NULL
+}
+
+void
+populate_capabilities (struct features_info_s *info, const char ***capabilities, size_t *num_capabilities)
+{
+  cap_value_t i;
+  *num_capabilities = 0;
+  for (i = 0;; i++)
+    {
+      char *v = cap_to_name (i);
+      if (v == NULL)
+        break;
+      char *endptr;
+      strtol (v, &endptr, 10);
+      if (endptr != v)
+        {
+          // Non-numeric or non-zero value encountered, break the loop
+          break;
+        }
+      (*num_capabilities)++;
+    }
+
+  *capabilities = xmalloc0 ((*num_capabilities + 1) * sizeof (const char *));
+  size_t index = 0;
+  for (i = 0;; i++)
+    {
+      char *v = cap_to_name (i);
+      if (v == NULL)
+        break;
+      char *endptr;
+      strtol (v, &endptr, 10);
+      if (endptr != v)
+        {
+          // Non-numeric or non-zero value encountered, break the loop
+          break;
+        }
+
+      // Convert capability name to uppercase
+      for (int j = 0; v[j] != '\0'; j++)
+        {
+          v[j] = toupper (v[j]);
+        }
+
+      (*capabilities)[index] = v;
+      index++;
+    }
+
+  (*capabilities)[index] = NULL; // Terminate the array with NULL
+  populate_array_field (&(info->linux.capabilities), *capabilities, *num_capabilities);
+}
+
+void
+retrieve_mount_options (struct features_info_s **info)
+{
+  // Retrieve mount options from wordlist
+  const struct propagation_flags_s *mount_options_list = get_mount_flags_from_wordlist ();
+
+  // Calculate the number of mount options
+  size_t num_mount_options = 0;
+  while (mount_options_list[num_mount_options].name != NULL)
+    {
+      num_mount_options++;
+    }
+
+  // Allocate memory for mount options in info struct
+  (*info)->mount_options = xmalloc0 (num_mount_options * sizeof (char *));
```

Comment:
make it NULL terminated: (*info)->mount_options = xmalloc0 ((num_mount_options+1) * sizeof (char *));